File: AgoraChat-ObjectiveC/AgoraChat/Code/CallKit/AgoraChatCallKitManager.m
//
//  AgoraChatCallKitManager.m
//  AgoraChat
//
//  Created by 冯钊 on 2022/4/11.
//  Copyright © 2022 easemob. All rights reserved.
//

#import "AgoraChatCallKitManager.h"

#import "AgoraChatCallConfig.h"
#import "AgoraChatCallManager.h"
#import "UserInfoStore.h"
#import "ConfInviteUsersViewController.h"
#import "AgoraMemberSelectViewController.h"

@import AVFAudio;
@import AgoraChat;

@interface AgoraChatCallKitManager() <AgoraChatCallDelegate>

@end

@implementation AgoraChatCallKitManager

+ (instancetype)shareManager
{
    static AgoraChatCallKitManager *share = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        share = [[self alloc] init];
    });
    return share;
}

- (instancetype)init
{
    if (self = [super init]) {
        AgoraChatCallConfig *config = [[AgoraChatCallConfig alloc] init];
        config.agoraAppId = @"15cb0d28b87b425ea613fc46f7c9f974";
        config.enableRTCTokenValidate = YES;
        config.enableIosCallKit = YES;

        [AgoraChatCallManager.sharedManager initWithConfig:config delegate:self];
    }
    return self;
}

- (void)updateAgoraUid:(NSInteger)agoraUid
{
    [AgoraChatCallManager.sharedManager getAgoraChatCallConfig].agoraUid = agoraUid;
}

- (void)audioCallToUser:(NSString *)userId
{
    AVAudioSessionRecordPermission permissionStatus = [[AVAudioSession sharedInstance] recordPermission];
    if (permissionStatus == AVAudioSessionRecordPermissionDenied) {
//        [EaseAlertController showErrorAlert:NSLocalizedString(@"needMicRight", nil)];
        return;
    }
    AgoraChatConversation *conversation = [AgoraChatClient.sharedClient.chatManager getConversationWithConvId:userId];
    NSString *msgId = conversation.latestMessage.messageId;
    AgoraChatUserInfo *info = [UserInfoStore.sharedInstance getUserInfoById:userId];
    if (info && (info.avatarUrl.length > 0 || info.nickname > 0)) {
        AgoraChatCallUser *user = [AgoraChatCallUser userWithNickName:info.nickname image:[NSURL URLWithString:info.avatarUrl]];
        [[AgoraChatCallManager.sharedManager getAgoraChatCallConfig] setUser:userId info:user];
    }
    [AgoraChatCallManager.sharedManager startSingleCallWithUId:userId type:AgoraChatCallType1v1Audio ext:nil completion:^(NSString * callId, AgoraChatCallError * aError) {
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 1000), dispatch_get_main_queue(), ^{
            [conversation loadMessagesStartFromId:msgId count:50 searchDirection:msgId ? AgoraChatMessageSearchDirectionDown : AgoraChatMessageSearchDirectionUp completion:^(NSArray *aMessages, AgoraChatError *aError) {
                if (aMessages.count) {
                    [self insertLocationCallRecord:aMessages];
                }
            }];
        });
    }];
}

- (void)videoCallToUser:(NSString *)userId
{
    AVAudioSessionRecordPermission permissionStatus = [[AVAudioSession sharedInstance] recordPermission];
    if (permissionStatus == AVAudioSessionRecordPermissionDenied) {
//        [EaseAlertController showErrorAlert:NSLocalizedString(@"needMicRight", nil)];
        return;
    }
    AVAuthorizationStatus authStatus = [AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeVideo];
    if (authStatus == AVAuthorizationStatusRestricted || authStatus ==AVAuthorizationStatusDenied) {
//        [EMAlertController showErrorAlert:NSLocalizedString(@"needCameraRight", nil)];
        return;
    }
    
    AgoraChatConversation *conversation = [AgoraChatClient.sharedClient.chatManager getConversationWithConvId:userId];
    NSString *msgId = conversation.latestMessage.messageId;
    AgoraChatUserInfo *info = [UserInfoStore.sharedInstance getUserInfoById:userId];
    if (info && (info.avatarUrl.length > 0 || info.nickname > 0)) {
        AgoraChatCallUser* user = [AgoraChatCallUser userWithNickName:info.nickname image:[NSURL URLWithString:info.avatarUrl]];
        [[AgoraChatCallManager.sharedManager getAgoraChatCallConfig] setUser:userId info:user];
    }
    [AgoraChatCallManager.sharedManager startSingleCallWithUId:userId type:AgoraChatCallType1v1Video ext:nil completion:^(NSString * callId, AgoraChatCallError * aError) {
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 1000), dispatch_get_main_queue(), ^{
            [conversation loadMessagesStartFromId:msgId count:50 searchDirection:msgId ? AgoraChatMessageSearchDirectionDown : AgoraChatMessageSearchDirectionUp completion:^(NSArray *aMessages, AgoraChatError *aError) {
                if (aMessages.count) {
                    [self insertLocationCallRecord:aMessages];
                }
            }];
        });
    }];
}

- (void)audioCallToGroup:(NSString *)groupId viewController:(UIViewController *)viewController
{
    ConfInviteUsersViewController *controller = [[ConfInviteUsersViewController alloc] initWithGroupId:groupId excludeUsers:@[AgoraChatClient.sharedClient.currentUsername]];
    controller.didSelectedUserList = ^(NSArray * _Nonnull aInviteUsers) {
        for (NSString *strId in aInviteUsers) {
            AgoraChatUserInfo *info = [UserInfoStore.sharedInstance getUserInfoById:strId];
            if (info && (info.avatarUrl.length > 0 || info.nickname > 0)) {
                AgoraChatCallUser *user = [AgoraChatCallUser userWithNickName:info.nickname image:[NSURL URLWithString:info.avatarUrl]];
                [[AgoraChatCallManager.sharedManager getAgoraChatCallConfig] setUser:strId info:user];
            }
        }
        [AgoraChatCallManager.sharedManager startInviteUsers:aInviteUsers groupId:groupId callType:AgoraChatCallTypeMultiAudio ext:@{
            @"groupId":groupId
        } completion:^(NSString * callId, AgoraChatCallError * aError) {
            
        }];
    };
    
    controller.modalPresentationStyle = UIModalPresentationPageSheet;
    [viewController presentViewController:controller animated:YES completion:nil];
}

- (void)videoCallToGroup:(NSString *)groupId viewController:(UIViewController *)viewController
{
    ConfInviteUsersViewController *controller = [[ConfInviteUsersViewController alloc] initWithGroupId:groupId excludeUsers:@[AgoraChatClient.sharedClient.currentUsername]];
    controller.didSelectedUserList = ^(NSArray * _Nonnull aInviteUsers) {
        for (NSString* strId in aInviteUsers) {
            AgoraChatUserInfo *info = [UserInfoStore.sharedInstance getUserInfoById:strId];
            if (info && (info.avatarUrl.length > 0 || info.nickname > 0)) {
                AgoraChatCallUser *user = [AgoraChatCallUser userWithNickName:info.nickname image:[NSURL URLWithString:info.avatarUrl]];
                [[AgoraChatCallManager.sharedManager getAgoraChatCallConfig] setUser:strId info:user];
            }
        }
        [AgoraChatCallManager.sharedManager startInviteUsers:aInviteUsers groupId:groupId callType:AgoraChatCallTypeMultiVideo ext:@{
            @"groupId": groupId
        } completion:^(NSString *callId, AgoraChatCallError *aError) {
            
        }];
    };
    
    controller.modalPresentationStyle = UIModalPresentationPageSheet;
    [viewController presentViewController:controller animated:YES completion:nil];
}

- (void)callDidEnd:(NSString*_Nonnull)aChannelName reason:(AgoraChatCallEndReason)aReason time:(int)aTm type:(AgoraChatCallType)aType
{
    NSString *msg = @"";
    switch (aReason) {
        case AgoraChatCallEndReasonAnswerOtherDevice:
            msg = NSLocalizedString(@"call.otherDeviceAnswer", nil);
            break;
        case AgoraChatCallEndReasonRefuseOtherDevice:
            msg = NSLocalizedString(@"call.otherDeviceRefuse", nil);
            break;
        case AgoraChatCallEndReasonBusy:
            msg = NSLocalizedString(@"call.busy", nil);
            break;
        case AgoraChatCallEndReasonRemoteRefuse:
            msg = NSLocalizedString(@"call.refuse", nil);
            break;
        case AgoraChatCallEndReasonNoResponse:
            msg = NSLocalizedString(@"call.noanswer", nil);
            break;
        case AgoraChatCallEndReasonHangup:
            break;
        default:
            break;
    }
    if (msg.length > 0) {
        [self showHint:msg];
    }
}

- (void)multiCallDidInvitingWithCurVC:(UIViewController *)vc callType:(AgoraChatCallType)callType excludeUsers:(NSArray<NSString *> *)users ext:(NSDictionary *)aExt
{
    NSString *groupId = nil;
    if (aExt) {
        groupId = [aExt objectForKey:@"groupId"];
    }
    if (!groupId || groupId.length <= 0) {
        return;
    }
    
    ConfInviteUsersViewController *confVC = [[ConfInviteUsersViewController alloc] initWithGroupId:groupId excludeUsers:users];
    confVC.didSelectedUserList = ^(NSArray * _Nonnull aInviteUsers) {
        for (NSString* strId in aInviteUsers) {
            AgoraChatUserInfo *info = [[UserInfoStore sharedInstance] getUserInfoById:strId];
            if (info && (info.avatarUrl.length > 0 || info.nickname.length > 0)) {
                AgoraChatCallUser *user = [AgoraChatCallUser userWithNickName:info.nickname image:[NSURL URLWithString:info.avatarUrl]];
                [[AgoraChatCallManager.sharedManager getAgoraChatCallConfig] setUser:strId info:user];
            }
        }
        [AgoraChatCallManager.sharedManager startInviteUsers:aInviteUsers groupId:groupId callType:AgoraChatCallTypeMultiAudio ext:aExt completion:nil];
    };
    confVC.modalPresentationStyle = UIModalPresentationPopover;
    [vc presentViewController:confVC animated:YES completion:nil];
}

- (void)callDidReceive:(AgoraChatCallType)aType inviter:(NSString*_Nonnull)username ext:(NSDictionary*_Nullable)aExt
{
    AgoraChatUserInfo* info = [[UserInfoStore sharedInstance] getUserInfoById:username];
    if (info && (info.avatarUrl.length > 0 || info.nickname.length > 0)) {
        AgoraChatCallUser *user = [AgoraChatCallUser userWithNickName:info.nickname image:[NSURL URLWithString:info.avatarUrl]];
        [[AgoraChatCallManager.sharedManager getAgoraChatCallConfig] setUser:username info:user];
    }
}

- (void)callDidOccurError:(AgoraChatCallError *_Nonnull)aError
{
    
}

- (void)callDidRequestRTCTokenForAppId:(NSString*_Nonnull)aAppId channelName:(NSString*_Nonnull)aChannelName account:(NSString*_Nonnull)aUserAccount uid:(NSInteger)aAgoraUid
{
    NSURLSessionConfiguration *config = [NSURLSessionConfiguration defaultSessionConfiguration];
    NSURLSession *session = [NSURLSession sessionWithConfiguration:config
                                                          delegate:nil
                                                     delegateQueue:[NSOperationQueue mainQueue]];
    
    NSString *strUrl = [NSString stringWithFormat:@"https://a41.chat.agora.io/token/rtc/channel/%@/agorauid/%@?userAccount=%@", aChannelName, @(aAgoraUid), aUserAccount];
    NSString *utf8Url = [strUrl stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLFragmentAllowedCharacterSet]];
    NSURL *url = [NSURL URLWithString:utf8Url];
    NSMutableURLRequest *urlReq = [[NSMutableURLRequest alloc] initWithURL:url];
    [urlReq setValue:[NSString stringWithFormat:@"Bearer %@",[AgoraChatClient sharedClient].accessUserToken ] forHTTPHeaderField:@"Authorization"];
    NSURLSessionDataTask *task = [session dataTaskWithRequest:urlReq completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (!data) {
            return;
        }
        NSDictionary *body = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:nil];
        NSLog(@"%@",body);
        if (!body) {
            return;
        }
        NSString *resCode = [body objectForKey:@"code"];
        if ([resCode isEqualToString:@"RES_OK"]) {
            NSString* rtcToken = [body objectForKey:@"accessToken"];
            [AgoraChatCallManager.sharedManager setRTCToken:rtcToken channelName:aChannelName uid:aAgoraUid];
        }
    }];

    [task resume];
}

- (void)remoteUserDidJoinChannel:( NSString*_Nonnull)aChannelName uid:(NSInteger)aUid username:(NSString*_Nullable)aUserName
{
    if (aUserName.length > 0) {
        AgoraChatUserInfo* userInfo = [[UserInfoStore sharedInstance] getUserInfoById:aUserName];
        if (userInfo && (userInfo.avatarUrl.length > 0 || userInfo.nickname.length > 0)) {
            AgoraChatCallUser* user = [AgoraChatCallUser userWithNickName:userInfo.nickname image:[NSURL URLWithString:userInfo.avatarUrl]];
            [[AgoraChatCallManager.sharedManager getAgoraChatCallConfig] setUser:aUserName info:user];
        }
    } else {
        [self _fetchUserMapsFromServer:aChannelName];
    }
}

- (void)callDidJoinChannel:(NSString*_Nonnull)aChannelName uid:(NSUInteger)aUid
{
    [self _fetchUserMapsFromServer:aChannelName];
}

- (void)insertLocationCallRecord:(NSArray *)messages
{
    [[NSNotificationCenter defaultCenter] postNotificationName:AGORA_CHAT_CALL_KIT_COMMMUNICATE_RECORD object:@{@"msg":messages}];//刷新页面
}

- (void)_fetchUserMapsFromServer:(NSString*)aChannelName
{
    // 这里设置映射表，设置头像，昵称
    NSURLSessionConfiguration *config = [NSURLSessionConfiguration defaultSessionConfiguration];
    NSURLSession *session = [NSURLSession sessionWithConfiguration:config
                                                          delegate:nil
                                                     delegateQueue:[NSOperationQueue mainQueue]];
    NSString *strUrl = [NSString stringWithFormat:@"https://a41.chat.agora.io/agora/channel/mapper?channelName=%@&userAccount=%@", aChannelName, [AgoraChatClient sharedClient].currentUsername];
    NSString*utf8Url = [strUrl stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLFragmentAllowedCharacterSet]];
    NSURL* url = [NSURL URLWithString:utf8Url];
    NSMutableURLRequest* urlReq = [[NSMutableURLRequest alloc] initWithURL:url];
    [urlReq setValue:[NSString stringWithFormat:@"Bearer %@",[AgoraChatClient sharedClient].accessUserToken ] forHTTPHeaderField:@"Authorization"];
    NSURLSessionDataTask *task = [session dataTaskWithRequest:urlReq completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (!data) {
            return;
        }
        NSDictionary *body = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:nil];
        NSLog(@"mapperBody:%@",body);
        if (!body) {
            return;
        }
        NSString *resCode = [body objectForKey:@"code"];
        if ([resCode isEqualToString:@"RES_OK"]) {
            NSString* channelName = [body objectForKey:@"channelName"];
            NSDictionary* result = [body objectForKey:@"result"];
            NSMutableDictionary<NSNumber*,NSString*>* users = [NSMutableDictionary dictionary];
            for (NSString *strId in result) {
                NSString *username = [result objectForKey:strId];
                NSNumber *uId = @([strId integerValue]);
                [users setObject:username forKey:uId];
                AgoraChatUserInfo* info = [[UserInfoStore sharedInstance] getUserInfoById:username];
                if (info && (info.avatarUrl.length > 0 || info.nickname.length > 0)) {
                    AgoraChatCallUser *user = [AgoraChatCallUser userWithNickName:info.nickname image:[NSURL URLWithString:info.avatarUrl]];
                    [[AgoraChatCallManager.sharedManager getAgoraChatCallConfig] setUser:username info:user];
                }
            }
            [AgoraChatCallManager.sharedManager setUsers:users channelName:channelName];
        }
    }];

    [task resume];
}

- (void)showHint:(NSString *)hint
{
    UIWindow *win = [self getKeyWindow];
    MBProgressHUD *hud = [MBProgressHUD showHUDAddedTo:win animated:YES];
    hud.userInteractionEnabled = NO;
    // Configure for text only and offset down
    hud.mode = MBProgressHUDModeText;
    hud.labelText = hint;
    hud.margin = 10.f;
    hud.yOffset = 180;
    hud.removeFromSuperViewOnHide = YES;
    [hud hide:YES afterDelay:2];
}

- (UIWindow *)getKeyWindow
{
    if (@available(iOS 13.0, *)) {
        for (UIWindowScene* scene in UIApplication.sharedApplication.connectedScenes) {
            if (@available(iOS 15.0, *)) {
                return scene.keyWindow;
            } else {
                for (UIWindow *window in scene.windows) {
                    if (window.isKeyWindow) {
                        return window;
                    }
                }
            }
        }
    } else {
        return [UIApplication sharedApplication].keyWindow;
    }
    return nil;
}

@end

File: AgoraChat-Swift/AgoraChat-Swift/code/CallKit/AgoraChatCallKitManager.swift
//
//  AgoraChatCallKitManager.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/19.
//

import UIKit

@objcMembers
class AgoraChatCallKitManager: NSObject {
    static let shared = AgoraChatCallKitManager()
    
    override init() {
        super.init()
        let config = AgoraChatCallConfig()
        config.agoraAppId = "15cb0d28b87b425ea613fc46f7c9f974"
        config.enableRTCTokenValidate = true
        config.enableIosCallKit = true
        AgoraChatCallManager.shared().initWith(config, delegate: self)
    }
    
    func update(agoraUid: UInt) {
        AgoraChatCallManager.shared().getAgoraChatCallConfig().agoraUid = agoraUid
    }
    
    func audioCall(toUser userId: String) {
        if AVAudioSession.sharedInstance().recordPermission == .denied {
            return
        }
        guard let conversation = AgoraChatClient.shared().chatManager?.getConversation(userId, type: .chat, createIfNotExist: true) else {
            return
        }
        let msgId: String? = conversation.latestMessage.messageId
        self.updateRTCUserInfo(userId: userId)
        AgoraChatCallManager.shared().startSingleCall(withUId: userId, type: .type1v1Audio, ext: nil) { callId, error in
            DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(1)) {
                conversation.loadMessagesStart(fromId: msgId, count: 50, searchDirection: msgId == nil ? .up : .down) { messages, error in
                    if let messages = messages, messages.count > 0 {
                        self.insertLocationCallRecord(messages: messages)
                    }
                }
            }
        }
    }

    func videoCall(toUser userId: String) {
        if AVAudioSession.sharedInstance().recordPermission == .denied {
            return
        }
        let authStatus = AVCaptureDevice.authorizationStatus(for: .video)
        if authStatus == .restricted || authStatus == .denied {
            return
        }
        guard let conversation = AgoraChatClient.shared().chatManager?.getConversation(userId, type: .chat, createIfNotExist: true) else {
            return
        }
        let msgId: String? = conversation.latestMessage.messageId
        self.updateRTCUserInfo(userId: userId)
        AgoraChatCallManager.shared().startSingleCall(withUId: userId, type: .type1v1Video, ext: nil) { callId, error in
            DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(1)) {
                conversation.loadMessagesStart(fromId: msgId, count: 50, searchDirection: msgId == nil ? .up : .down) { messages, error in
                    if let messages = messages, messages.count > 0 {
                        self.insertLocationCallRecord(messages: messages)
                    }
                }
            }
        }
    }
    
    func audioCall(toGroup groupId: String, viewController: UIViewController) {
        let vc = ConfInviteUsersViewController(groupId: groupId, excludeUsers: [AgoraChatClient.shared().currentUsername!])
        vc.didSelectedUserListHandle = { inviteUsers in
            for userId in inviteUsers {
                self.updateRTCUserInfo(userId: userId)
            }
            AgoraChatCallManager.shared().startInviteUsers(inviteUsers, groupId: groupId, callType: .typeMultiAudio, ext: [
                "groupId": groupId
            ], completion: nil)
        }
        vc.modalPresentationStyle = .pageSheet
        viewController.present(vc, animated: true)
    }

    func videoCall(toGroup groupId: String, viewController: UIViewController) {
        let vc = ConfInviteUsersViewController(groupId: groupId, excludeUsers: [AgoraChatClient.shared().currentUsername!])
        vc.didSelectedUserListHandle = { inviteUsers in
            for userId in inviteUsers {
                self.updateRTCUserInfo(userId: userId)
            }
            AgoraChatCallManager.shared().startInviteUsers(inviteUsers, groupId: groupId, callType: .typeMultiVideo, ext: [
                "groupId": groupId
            ], completion: nil)
        }
        vc.modalPresentationStyle = .pageSheet
        viewController.present(vc, animated: true)
    }
    
    private func insertLocationCallRecord(messages: [AgoraChatMessage]) {
        NotificationCenter.default.post(name: CallKitRecordMessageNotification, object: [
            "msg": messages
        ])
    }
    
    private func updateRTCUserInfo(userId: String) {
        if let info = UserInfoStore.shared.getUserInfo(userId: userId) {
            let user = AgoraChatCallUser(nickName: info.nickname, image: info.avatarUrl == nil ? nil : URL(string: info.avatarUrl!))
            AgoraChatCallManager.shared().getAgoraChatCallConfig().setUser(userId, info: user)
        }
    }
    
    private func fetchUserMapsFromServer(channelName: String) {
        let config = URLSessionConfiguration.default
        let session = URLSession(configuration: config)
        let strUrl = "http://a41.easemob.com/agora/channel/mapper?channelName=\(channelName)&userAccount=\(AgoraChatClient.shared().currentUsername!)"
        guard let utf8Url = strUrl.addingPercentEncoding(withAllowedCharacters: .urlFragmentAllowed) else {
            return
        }
        guard let url = URL(string: utf8Url) else {
            return
        }
        var request = URLRequest(url: url)
        request.setValue("Bearer \(AgoraChatClient.shared().accessUserToken ?? "")", forHTTPHeaderField: "Authorization")
        let task = session.dataTask(with: request) { data, response, error in
            guard let data = data else {
                return
            }
            guard let body = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                return
            }
            let resCode = body["code"] as? String
            if resCode == "RES_OK", let channelName = body["channelName"] as? String {
                let result = body["result"] as? [String: String]
                var users: [NSNumber: String] = [:]
                if let result = result {
                    for (strId, username) in result {
                        if let uid = Int(strId) {
                            users[NSNumber(value: uid)] = username
                        }
                        self.updateRTCUserInfo(userId: username)
                    }
                }
                DispatchQueue.main.async {
                    AgoraChatCallManager.shared().setUsers(users, channelName: channelName)
                }
            }
        }
        task.resume()
    }
}

extension AgoraChatCallKitManager: AgoraChatCallDelegate {
    func callDidEnd(_ aChannelName: String, reason aReason: AgoraChatCallEndReason, time aTm: Int32, type aType: AgoraChatCallType) {
        switch aReason {
        case .answerOtherDevice:
            self.showHint("Other devices connected".localized)
        case .refuseOtherDevice:
            self.showHint("Other devices declined".localized)
        case .busy:
            self.showHint("The line is busy".localized)
        case .remoteRefuse:
            self.showHint("Request declined".localized)
        case .noResponse:
            self.showHint("No response".localized)
        default:
            break
        }
    }
    
    func multiCallDidInviting(withCurVC vc: UIViewController, callType: AgoraChatCallType, excludeUsers users: [String]?, ext aExt: [AnyHashable : Any]?) {
        guard let groupId = aExt?["groupId"] as? String, groupId.count > 0 else {
            return
        }
        let confVC = ConfInviteUsersViewController(groupId: groupId, excludeUsers: users ?? [])
        confVC.didSelectedUserListHandle = { inviteUsers in
            for userId in inviteUsers {
                self.updateRTCUserInfo(userId: userId)
            }
            AgoraChatCallManager.shared().startInviteUsers(inviteUsers, groupId: groupId, callType: callType, ext: aExt, completion: nil)
        }
        confVC.modalPresentationStyle = .popover
        vc.present(confVC, animated: true)
    }
    
    func callDidReceive(_ aType: AgoraChatCallType, inviter user: String, ext aExt: [AnyHashable : Any]?) {
        self.updateRTCUserInfo(userId: user)
    }
    
    func callDidOccurError(_ aError: AgoraChatCallError) {
        
    }
    
    func callDidRequestRTCToken(forAppId aAppId: String, channelName aChannelName: String, account aUserAccount: String, uid aAgoraUid: Int) {
        let config = URLSessionConfiguration.default
        let session = URLSession(configuration: config)
        let strUrl = "http://a41.easemob.com/token/rtc/channel/\(aChannelName)/agorauid/\(aAgoraUid)?userAccount=\(aUserAccount)"
        guard let utf8Url = strUrl.addingPercentEncoding(withAllowedCharacters: .urlFragmentAllowed) else {
            return
        }
        guard let url = URL(string: utf8Url) else {
            return
        }
        var request = URLRequest(url: url)
        request.setValue("Bearer \(AgoraChatClient.shared().accessUserToken ?? "")", forHTTPHeaderField: "Authorization")
        let task = session.dataTask(with: request) { data, response, error in
            guard let data = data else {
                return
            }
            guard let body = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                return
            }
            let resCode = body["code"] as? String
            if resCode == "RES_OK", let rtcToken = body["accessToken"] as? String {
                AgoraChatCallManager.shared().setRTCToken(rtcToken, channelName: aChannelName, uid: UInt(aAgoraUid))
            }
        }
        task.resume()
    }
    
    func remoteUserDidJoinChannel(_ aChannelName: String, uid aUid: Int, username aUserName: String?) {
        if let userName = aUserName, userName.count > 0  {
            self.updateRTCUserInfo(userId: userName)
        } else {
            self.fetchUserMapsFromServer(channelName: aChannelName)
        }
    }
    
    func callDidJoinChannel(_ aChannelName: String, uid aUid: UInt) {
        self.fetchUserMapsFromServer(channelName: aChannelName)
    }
}
